pkg/kubectl/cmd/drain.go

```diff
@@ -195,17 +195,55 @@ func (o *DrainOptions) RunDrain() error {
 	return nil
 }
 
-// GetPodsForDeletion returns all the pods we're going to delete.  If there are
+// getPodsForDeletion returns all the pods we're going to delete.  If there are
 // any unmanaged pods and the user didn't pass --force, we return that list in
 // an error.
-func (o *DrainOptions) GetPodsForDeletion() ([]api.Pod, error) {
-	pods := []api.Pod{}
-	podList, err := o.client.Pods(api.NamespaceAll).List(api.ListOptions{FieldSelector: fields.SelectorFromSet(fields.Set{"spec.nodeName": o.nodeInfo.Name})})
+func (o *DrainOptions) getPodsForDeletion() ([]api.Pod, error) {
+	pods, unreplicatedPodNames, daemonSetPodNames, err := GetPodsForDeletionOnNodeDrain(
+		o.client,
+		o.nodeInfo.Name,
+		o.factory.Decoder(true),
+		o.Force,
+		o.IgnoreDaemonsets,
+	)
 	if err != nil {
-		return pods, err
+		return []api.Pod{}, err
+	}
+
+	daemonSetErrors := !o.IgnoreDaemonsets && len(daemonSetPodNames) > 0
+	unreplicatedErrors := !o.Force && len(unreplicatedPodNames) > 0
+
+	switch {
+	case daemonSetErrors && unreplicatedErrors:
+		return []api.Pod{}, errors.New(unmanagedMsg(unreplicatedPodNames, daemonSetPodNames, true))
+	case daemonSetErrors && !unreplicatedErrors:
+		return []api.Pod{}, errors.New(unmanagedMsg([]string{}, daemonSetPodNames, true))
+	case unreplicatedErrors && !daemonSetErrors:
+		return []api.Pod{}, errors.New(unmanagedMsg(unreplicatedPodNames, []string{}, true))
+	}
+
+	if len(unreplicatedPodNames) > 0 {
+		fmt.Fprintf(o.out, "WARNING: About to delete these %s\n", unmanagedMsg(unreplicatedPodNames, []string{}, false))
+	}
+	if len(daemonSetPodNames) > 0 {
+		fmt.Fprintf(o.out, "WARNING: Skipping %s\n", unmanagedMsg([]string{}, daemonSetPodNames, false))
+	}
+
+	return pods, nil
+}
+
+// GetPodsForDeletionOnNodeDrain returns pods that should be deleted on node drain as well as some extra information
+// about possibly problematic pods (unreplicated and deamon sets).
+func GetPodsForDeletionOnNodeDrain(client *client.Client, nodename string, decoder runtime.Decoder, removeUnderplicated bool,
```

Comment:
The whole purpose of this refactoring is to make it public :).

Other comments applied.